lib/puppetserver/ca/cli.rb

```diff
@@ -0,0 +1,76 @@
+require 'optparse'
+require 'puppetserver/ca/version'
+require 'puppetserver/ca/setup_command'
+require 'puppetserver/ca/logger'
+
+module Puppetserver
+  module Ca
+    class Cli
+      VALID_COMMANDS = {'setup' => SetupCommand}
+
+      def self.run!(cli_args = ARGV, out = STDOUT, err = STDERR)
+        logger = Puppetserver::Ca::Logger.new(:info, out, err)
+        parser, general_options, unparsed = parse_general_inputs(cli_args)
+
+        if general_options['version']
+          logger.inform Puppetserver::Ca::VERSION
+          return 0
+        end
+
+        subcommand = VALID_COMMANDS[unparsed.first]
```

Comment:
Do we want to return an error if it's an unrecognized subcommand, or just return help message?

Reply:
We should return an exit code >0 for an unrecognized subcommand (which we do below). It could be nice, but not entirely necessary to print an error message before the help text.

Reply:
I think we want to exit 1 (denoting an error) and print the help to stderr. We should probably also print out that we did not recognize a subcommand if one was passed. I'll add that around line 42 where we handle missing subcommands.